src/termination.ml

```diff
@@ -0,0 +1,378 @@
+     else
+       let (x, b) = Bindlib.unbind b in
+       out "<arrow>@.<var>v_%s</var>@." (Bindlib.name_of x);
+       out "<type>@.%a</type>@.<type>@.%a</type>@.</arrow>"
+         (print_type i s) a (print_type i s) b
+       
+(** [print_rule oc s r] outputs the rule declaration corresponding [r] (on the
+    symbol [s]), to the output channel [oc]. *)
+let print_rule : Format.formatter -> int -> sym -> rule -> unit =
+  fun oc i s r ->
+  (* Print the rewriting rule. *)
+  let lhs = Basics.add_args (Symb(s,Qualified)) r.lhs in
+  Format.fprintf oc "<rule>@.<lhs>@.%a</lhs>@." (print_term i s.sym_name) lhs;
+  let rhs = Basics.term_of_rhs r in
+  Format.fprintf oc "<rhs>@.%a</rhs>@.</rule>@." (print_term i s.sym_name) rhs
+  
+(** [print_tl_rule] is identical to [print_rule] but for type-level rule  *)
+let print_tl_rule : Format.formatter -> int -> sym -> rule -> unit =
+  fun oc i s r ->
+  (* Print the type level rewriting rule. *)
+  let lhs = Basics.add_args (Symb(s,Qualified)) r.lhs in
+  Format.fprintf oc "<typeLevelRule>@.<TLlhs>@.%a</TLlhs>@."
+    (print_type i s.sym_name) lhs;
+  let rhs = Basics.term_of_rhs r in
+  Format.fprintf oc "<TLrhs>@.%a</TLrhs>@.</typeLevelRule>@."
+    (print_type i s.sym_name) rhs
+  
+(** [get_vars s r] returns the list of variable of used in the rule [r],
+    in the form of a couple containing the name of the variable and its type,
+    inferred by the solver. *)
+let get_vars : sym -> rule -> (string * Terms.term) list = fun s r ->
+  (*check_rule s r;*)
+  let rule_ctx : tvar option array = Array.make (Array.length r.vars) None in
+  let var_list : tvar list ref = ref [] in
+  let rec subst_patt v t =
+    match t with
+    | Type
+    | Kind
+    | TEnv (_, _)
+    | Meta (_, _)
+    | TRef _
+    | Wild
+    | Prod (_, _)
+    | Vari _              -> assert false
+    | Symb (s, p)         -> Symb(s, p)
+    | Abst (t1, b)        ->
+       let (x,t2) = Bindlib.unbind b in
+       Abst(
+           subst_patt v t1,
+           Bindlib.unbox (Bindlib.bind_var x (lift (subst_patt v t2)))
+         )
+    | Appl (t1, t2)        -> Appl(subst_patt v t1, subst_patt v t2)
+    | Patt (None, x, a)    ->
+       let v_i = Bindlib.new_var mkfree x in
+       var_list := v_i :: !var_list;
+       Array.fold_left (fun acc t -> Appl(acc,t))
+         (Vari(v_i)) a
+    | Patt (Some(i), x, a) ->
+       if v.(i) = None
+       then
+         (let v_i = Bindlib.new_var mkfree x in
+          var_list := v_i :: !var_list;
+          v.(i) <- Some(v_i));
+       let v_i =
+         match v.(i) with
+         |Some(vi) -> vi
+         |None -> assert false
+       in
+       Array.fold_left (fun acc t -> Appl(acc,t)) (Vari(v_i)) a
+  in
+  let lhs =
+    List.fold_left
+      (fun t h -> Appl(t,subst_patt rule_ctx (unfold h)))
+      (Symb(s,Nothing)) r.lhs
+  in
+  let ctx =
+    List.fold_left (fun l x -> (x,Meta(fresh_meta Type 0,[||]))::l) [] !var_list
+  in
+  let (_,l) = Typing.infer ctx lhs in
```

Comment:
Couldn't we avoid doing type inference *again*? Because it is already done by the SR algorithm! We would just need to store the types of variables somewhere in the type rule.